File: src/schema.rs
//! Serde types matching the OpenAI-compatible chat-completions shape that the
//! MTPLX server accepts. Kept narrow on purpose — only the fields we actually
//! send or read.

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// A single chat message. `role` is one of "system", "user", "assistant", "tool".
/// `content` may be a plain string or null when the assistant is purely calling
/// tools. We keep it as `Option<String>` for symmetry with OpenAI.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatMessage {
    pub role: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<Vec<ToolCall>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
}

impl ChatMessage {
    pub fn system(s: impl Into<String>) -> Self {
        Self {
            role: "system".into(),
            content: Some(s.into()),
            tool_calls: None,
            tool_call_id: None,
        }
    }
    pub fn user(s: impl Into<String>) -> Self {
        Self {
            role: "user".into(),
            content: Some(s.into()),
            tool_calls: None,
            tool_call_id: None,
        }
    }
    pub fn assistant_text(s: impl Into<String>) -> Self {
        Self {
            role: "assistant".into(),
            content: Some(s.into()),
            tool_calls: None,
            tool_call_id: None,
        }
    }
    pub fn assistant_tool_calls(content: Option<String>, calls: Vec<ToolCall>) -> Self {
        Self {
            role: "assistant".into(),
            content,
            tool_calls: Some(calls),
            tool_call_id: None,
        }
    }
    pub fn tool_result(id: impl Into<String>, body: impl Into<String>) -> Self {
        Self {
            role: "tool".into(),
            content: Some(body.into()),
            tool_calls: None,
            tool_call_id: Some(id.into()),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCall {
    pub id: String,
    #[serde(rename = "type", default = "default_type")]
    #[allow(dead_code)]
    pub kind: String,
    pub function: FunctionCall,
}

fn default_type() -> String {
    "function".to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FunctionCall {
    pub name: String,
    /// JSON-encoded string per OpenAI convention.
    pub arguments: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct ChatRequest<'a> {
    pub model: &'a str,
    pub messages: &'a [ChatMessage],
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tools: Option<&'a [Value]>,
    pub stream: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_p: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_k: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub chat_template_kwargs: Option<Value>,
}

/// Streaming delta envelope. We only decode the fields we use.
#[derive(Debug, Deserialize)]
pub struct StreamChunk {
    #[serde(default)]
    pub choices: Vec<StreamChoice>,
    #[serde(default)]
    pub usage: Option<Usage>,
}

#[derive(Debug, Deserialize)]
pub struct StreamChoice {
    #[serde(default)]
    pub delta: Delta,
    #[serde(default)]
    pub finish_reason: Option<String>,
}

#[derive(Debug, Default, Deserialize)]
pub struct Delta {
    #[serde(default)]
    #[allow(dead_code)]
    pub role: Option<String>,
    #[serde(default)]
    pub content: Option<String>,
    #[serde(default)]
    pub tool_calls: Option<Vec<ToolCallDelta>>,
}

#[derive(Debug, Deserialize)]
pub struct ToolCallDelta {
    pub index: u32,
    #[serde(default)]
    pub id: Option<String>,
    #[serde(default, rename = "type")]
    #[allow(dead_code)]
    pub kind: Option<String>,
    #[serde(default)]
    pub function: Option<FunctionCallDelta>,
}

#[derive(Debug, Deserialize)]
pub struct FunctionCallDelta {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub arguments: Option<String>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct Usage {
    #[serde(default)]
    pub prompt_tokens: Option<u32>,
    #[serde(default)]
    pub completion_tokens: Option<u32>,
    #[serde(default)]
    #[allow(dead_code)]
    pub total_tokens: Option<u32>,
}

File: src/schema.ts
// OpenAI-compatible chat schema. Matches what MTPLX accepts/emits over
// its /v1/chat/completions endpoint with stream=true. Field names mirror
// hip's Rust ChatMessage and ToolCall structs so the wire format stays
// identical.

export type Role = 'system' | 'user' | 'assistant' | 'tool';

export interface FunctionCall {
  name: string;
  /** JSON-encoded arguments string (per OpenAI spec, not an object). */
  arguments: string;
}

export interface ToolCall {
  id: string;
  /** "function" - kept verbose for OpenAI parity. */
  type: 'function';
  function: FunctionCall;
}

export interface ChatMessage {
  role: Role;
  content: string | null;
  /** Populated on assistant messages that asked for tool calls. */
  tool_calls?: ToolCall[];
  /** Required on role=tool messages to thread back to the originating call. */
  tool_call_id?: string;
  /** Optional name (rare; mostly for role=tool replies). */
  name?: string;
}

export interface ToolSpec {
  type: 'function';
  function: {
    name: string;
    description?: string;
    parameters: Record<string, unknown>;
  };
}

export interface SSEDelta {
  content?: string;
  role?: Role;
  tool_calls?: Array<{
    index: number;
    id?: string;
    type?: 'function';
    function?: { name?: string; arguments?: string };
  }>;
}

export interface SSEChoice {
  index: number;
  delta: SSEDelta;
  finish_reason?: string;
}

export interface SSEChunk {
  id?: string;
  choices: SSEChoice[];
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
    total_tokens?: number;
  };
  error?: unknown;
}

export interface CompletionResult {
  content: string;
  tool_calls: ToolCall[];
  finish_reason?: string;
  usage?: SSEChunk['usage'];
  ttftMs?: number;
  totalMs?: number;
}

export function systemMessage(text: string): ChatMessage {
  return { role: 'system', content: text };
}
export function userMessage(text: string): ChatMessage {
  return { role: 'user', content: text };
}
export function assistantText(text: string): ChatMessage {
  return { role: 'assistant', content: text };
}
export function assistantToolCalls(text: string, calls: ToolCall[]): ChatMessage {
  return { role: 'assistant', content: text || null, tool_calls: calls };
}
export function toolResultMessage(call_id: string, name: string, text: string): ChatMessage {
  return { role: 'tool', tool_call_id: call_id, name, content: text };
}
